server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1731,31 +1732,39 @@ public void close(String reason, boolean flushEngine, Executor closeExecutor, Ac
                 checkAndCallWaitForEngineOrClosedShardListeners();
             } finally {
                 final Engine engine = this.currentEngineReference.getAndSet(null);
-                closeExecutor.execute(ActionRunnable.run(closeListener, new CheckedRunnable<>() {
-                    @Override
-                    public void run() throws Exception {
-                        try {
-                            if (engine != null && flushEngine) {
-                                engine.flushAndClose();
-                            }
-                        } finally {
+                closeExecutor.execute(new AbstractRunnable() {
+
+                    protected void doRun() throws Exception {
+                        var cleanUpAndClose = ActionListener.runBefore(closeListener, () -> {
                             // playing safe here and close the engine even if the above succeeds - close can be called multiple times
                             // Also closing refreshListeners to prevent us from accumulating any more listeners
+                            // TODO Consider closing the engine asynchronously, but since it should be already be closed,
+                            // a sync call should complete immediately
                             IOUtils.close(
-                                engine,
+                                engine != null ? () -> engine.close(ActionListener.noop()) : null,
```

Comment:
We can't use `ActionListener.noop()` here, we need to wait for `engine.close()` to complete before closing the rest of the resources and completing the outer listener. I know it makes no difference here today but the reason I'd like to split this change out first is to get all the callers into the right async state first.